soc/st/stm32/common/stm32_wkup_pins.c

```diff
@@ -133,7 +131,31 @@ struct wkup_pin_cfg_t {
  * @brief LookUp Table to store LL_PWR_WAKEUP_PINx for each wake-up pin.
  */
 static const uint32_t table_wakeup_pins[PWR_STM32_MAX_NB_WKUP_PINS + 1] = {
-	LISTIFY(PWR_STM32_MAX_NB_WKUP_PINS, PWR_STM32_WKUP_PIN_DEF, (,))};
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 1 && defined(LL_PWR_WAKEUP_PIN1)
+	[1] = LL_PWR_WAKEUP_PIN1,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 2 && defined(LL_PWR_WAKEUP_PIN2)
+	[2] = LL_PWR_WAKEUP_PIN2,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 3 && defined(LL_PWR_WAKEUP_PIN3)
+	[3] = LL_PWR_WAKEUP_PIN3,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 4 && defined(LL_PWR_WAKEUP_PIN4)
+	[4] = LL_PWR_WAKEUP_PIN4,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 5 && defined(LL_PWR_WAKEUP_PIN5)
+	[5] = LL_PWR_WAKEUP_PIN5,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 6 && defined(LL_PWR_WAKEUP_PIN6)
+	[6] = LL_PWR_WAKEUP_PIN6,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 7 && defined(LL_PWR_WAKEUP_PIN7)
+	[7] = LL_PWR_WAKEUP_PIN7,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 8 && defined(LL_PWR_WAKEUP_PIN8)
+	[8] = LL_PWR_WAKEUP_PIN8,
+#endif
+};
```

Comment:
`IF_ENABLED` can be used inside macros as replacement for `#if defined()`, which allows us to put the check in a macro that can be `LISTIFY`'ed.

For the off-by-one, I think it's reasonable to have a dummy value at index 0, before the `LISTIFY`.

Rough idea of what this could look like:
```c
/* Keep PWR_STM32_WKUP_PIN_DEF defined (although it could use _CONCAT instead) */

/* To be discussed: are discontinuous ranges a potential issue? (e.g., PIN3 exists but not PIN2)
 * If yes, COND_CODE_1 should be used instead.
 */
#define PIN_IF_EXISTS(x) \
    IF_ENABLED(PWR_STM32_WKUP_PIN_DEF(x), (PWR_STM32_WKUP_PIN_DEF(x)))

/* LISTIFY is [0..n-1] but we care about [1..n]
#define LISTIFY_WKUP_PIN_HELPER(x) \
        PIN_IF_EXISTS(UTIL_INC(x))
```

then the whole array can be replaced with a simple:
```suggestion
  0, /* Dummy entry - PIN0 does not exist */
  LISTIFY(PWR_STM32_MAX_NB_WKUP_PINS, LISTIFY_WKUP_PIN_HELPER, (,))
};
```

Reply:
Thank you for looking at this!

Unless I'm missing something, that still doesn't populate a valid array member at `table_wakeup_pins[PWR_STM32_MAX_NB_WKUP_PINS]`, because `LISTIFY` ranges over [0, limit). E.g., with `PWR_STM32_MAX_NB_WKUP_PINS == 5`, you'll get an array with `{0, LL_PWR_WAKEUP_PIN1, LL_PWR_WAKEUP_PIN2, LL_PWR_WAKEUP_PIN3, LL_PWR_WAKEUP_PIN4}`.

> are discontinuous ranges a potential issue? (e.g., PIN3 exists but not PIN2)

Just based on one of the STM32 LL headers (`stm32g0xx_ll_pwr.h`), it does appear to be possible to have a STM32G0 with WKUP4 but not WKUP3.

Reply:
Ah, sorry, I missed `UTIL_INC`! I'll try this.

Reply:
> > are discontinuous ranges a potential issue? (e.g., PIN3 exists but not PIN2)
> 
> Just based on one of the STM32 LL headers (`stm32g0xx_ll_pwr.h`), it does appear to be possible to have a STM32G0 with WKUP4 but not WKUP3.

Indeed. This also appears to be the case for STM32WB series.

Since the corresponding `wkup-pin` DTS nodes should not be active on these series, it sounds safe to have the corresponding indexes in the array be zero: in this case, the following using `COND_CODE_1` (as I suggested) should work:
```c
#define PIN_IF_EXISTS(x) \
    COND_CODE_1(PWR_STM32_WKUP_PIN_DEF(x),
         (PWR_STM32_WKUP_PIN_DEF(x)),
         (0)
    )
```